### PR TITLE
refactor input specs and shutdown sequence

### PR DESCRIPTION
**Pull request by colinsurprenant**
this also requires the new logstash-core 1.5.0.rc3

### CODE REVIEW

File: lib/logstash/devutils/rspec/logstash_helpers.rb [hunk 1]
```diff
@@ -36,32 +37,45 @@ def sample(sample_event, &block)
...
         pipeline.flush_filters(:final => true) do |e|
           results << e unless e.cancelled?
         end
 
-        results
+        results.select{|e| !e.cancelled?}
       end
 
-      subject { results.length > 1 ? results: results.first }
+      subject { results.length > 1 ? results : results.first }
```
Comment (purbon) [inline lib/logstash/devutils/rspec/logstash_helpers.rb]: In order to be uniform, is good to remove the space after the brackets in here, isn't?

File: lib/logstash/devutils/rspec/spec_helper.rb [hunk 2]
```diff
@@ -43,6 +43,7 @@ def []=(str, value)
 end
 
 RSpec.configure do |config|
+  config.include LogStashHelper
```
Comment (purbon) [inline lib/logstash/devutils/rspec/spec_helper.rb]: I tend to prefer the syntax where in the Helper include, you add also the extend call ... I find this way more clear than this one in here that basically need two calls, but is my experience, do we want to keep this logic within the module?

File: lib/logstash/devutils/rspec/logstash_helpers.rb [hunk 3]
```diff
@@ -36,32 +37,45 @@ def sample(sample_event, &block)
...
-          pipeline.filter(e) {|new_event| results << new_event }
+          pipeline.filter(e) {|new_event| results << new_event}
         end
 
         pipeline.flush_filters(:final => true) do |e|
           results << e unless e.cancelled?
         end
 
-        results
+        results.select{|e| !e.cancelled?}
```
Comment (purbon) [inline lib/logstash/devutils/rspec/logstash_helpers.rb]: Why do we need this filter in here? aren't we doing the same work just before in the flush filter call?

File: lib/logstash/devutils/rspec/spec_helper.rb (hunk 2 above)
Comment (colinsurprenant) [inline lib/logstash/devutils/rspec/spec_helper.rb]: both extend and include are required for now because of the "legacy" dsl addons.

Comment (purbon) [inline lib/logstash/devutils/rspec/spec_helper.rb]: sure, but why not doing this in the module, using the included? hook

File: lib/logstash/devutils/rspec/logstash_helpers.rb (hunk 3 above)
Comment (colinsurprenant) [inline lib/logstash/devutils/rspec/logstash_helpers.rb]: no, flush only takes care of  the flushed events if any. we could "optimize" and rewrite the flushing to be like
```ruby
         pipeline.flush_filters(:final => true) do |e|
           results << e
         end
```
but in the end it does not fix anything,

Comment (purbon) [inline lib/logstash/devutils/rspec/logstash_helpers.rb]: ok, no problem.

File: lib/logstash/devutils/rspec/spec_helper.rb (hunk 2 above)
Comment (colinsurprenant) [inline lib/logstash/devutils/rspec/spec_helper.rb]: Hm, I think it's more explicit to do both here as it clearly state the intention of both extending and including.

Comment (purbon) [inline lib/logstash/devutils/rspec/spec_helper.rb]: No problem, just a matter of style in here --> as I said *I tend to prefer the syntax*, it does not mean it should be like this for sure :-)

**Pull request merged**